File: algorithms/algorithms-cplusplus/coding_interviews_2/Question_0022_KthNodeFromEnd.cc

// Definition for singly-linked list.
struct ListNode {
  int val;
  ListNode *next;
  ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
 public:
  ListNode* FindKthToTail(ListNode* pListHead, unsigned int k) {
    if (k <= 0) {
      return NULL;
    }
    ListNode *p = pListHead;
    while (p && k > 0) {
      // 由于是unsigned int，在while条件中判断会出错
      k--;
      p = p->next;
    }
    if (k > 0) {
      return NULL;
    }
    while (p) {
      pListHead = pListHead->next;
      p = p->next;
    }
    return pListHead;
  }
};

File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/coding_interviews_2/Question_0022_KthNodeFromEnd.java
package com.lanceolata.coding.algorithms.java.coding_interviews_2;

public class Question_0022_KthNodeFromEnd {
    // Definition for singly-linked list.
    public class ListNode {
        int val;
        ListNode next;
        ListNode(int x) {
            val = x;
            next = null;
        }
    }

    public ListNode FindKthToTail(ListNode head, int k) {
        if (head == null || k <= 0) {
            return null;
        }
        ListNode p = head;
        while (p != null && k-- > 0) {
            p = p.next;
        }
        if (k > 0) {
            return null;
        }
        while (p != null) {
            head = head.next;
            p = p.next;
        }
        return head;
    }
}
